### PR TITLE
Env changes to decouple contract instance from source (CAP-46-02)

### PR DESCRIPTION
**Pull request by dmkozh**
### What

This change implements separation WASM sources as per updated CAP-46-02. It also uses the update host function call interface.

### Why

This allows to efficiently the reuse the contract sources, so that it's possible to have multiple contract instances that share the same single ledger entry with WASM code.

### Known limitations

This temporarily removes the host fn that allows to create a contract from contract on behalf of a `ed25519` public key (they weren't implemented anyway). We may bring this back in a bit different form if we change the authorization.

### CODE REVIEW

File: soroban-env-host/src/host.rs [hunk 1]
```diff
@@ -1058,6 +1032,57 @@ impl Host {
...
+            self.with_mut_storage(|storage| {
+                let data = LedgerEntryData::ContractCode(ContractCodeEntry {
+                    hash: Hash(hash_bytes),
+                    code: args.code,
+                    ext: ExtensionPoint::V0,
+                });
+                storage.put(
+                    &code_key,
+                    &LedgerEntry {
+                        last_modified_ledger_seq: 0,
```
Comment (sisuresh) [inline soroban-env-host/src/host.rs:1075]: I think setting this to 0 is fine because `LedgerTxn:commit` will set last modified later on, but I think we should add a comment explaining this? We do this in a few places so maybe a helper that takes just a `LedgerEntryData` and returns a `LedgerEntry` is a good idea. If you'd like, I can open an issue for this instead since it's unrelated to this PR.

File: soroban-env-common/src/env.rs [hunk 2]
```diff
@@ -304,21 +304,13 @@ macro_rules! call_macro_with_all_host_functions {
                 {"0", fn has_contract_data(k:RawVal) -> RawVal}
                 {"1", fn get_contract_data(k:RawVal) -> RawVal}
                 {"2", fn del_contract_data(k:RawVal) -> RawVal}
-                {"3", fn create_contract_from_ed25519(v: Object, salt: Object, key: Object, sig: Object) -> Object}
```
Comment (sisuresh) [inline soroban-env-common/src/env.rs:307]: I just remembered something - The main reason for allowing contracts to create other contracts using ed25519 was to allow for factory contracts that take the parameters as input to the contract function. This doesn’t work well with `create_contract_from_contract` because another party can frontrun with the same salt but different wasm_hash. I know this scenario is an edge case and the contract creator can always verify the end result before sharing the contract id, but I’m curious what you think about this.

File: soroban-env-host/src/test/util.rs [hunk 3]
```diff
@@ -145,4 +134,39 @@ impl Host {
...
+        for val in v {
+            res.push(val.try_into_val(self).unwrap());
+        }
+        res.try_into().unwrap()
+    }
+
+    // Registers a contract with provided WASM source and returns the registered
+    // contract ID.
+    // This relies on the host to have no footprint enforcement.
+    pub(crate) fn register_test_contract_wasm(
```
Comment (sisuresh) [inline soroban-env-host/src/test/util.rs:149]: I see that this was moved from host.rs, but I don't see `register_test_contract_token` anywhere. I'm guessing it was removed but not added back in because only the sdk uses it? Or do you have a pending sdk change that no longer requires it?

**Review by jayz22 — COMMENTED** [review]
I've done a review on the implementation while trusting the design (I'm still catching up my understanding on the design). 
Just a few trivial/straightforward comments. Overall great work, really nice to see the contract creation/deployment workflows getting streamlined and cleaned up!

File: soroban-env-host/src/host.rs (hunk 1 above)
Comment (dmkozh) [reply soroban-env-host/src/host.rs]: While it's not strictly relevant, most of the use cases for this are introduced/changed in this PR, so adding a helper makes sense.

**Review by dmkozh — COMMENTED** [review]
PTAL

File: soroban-env-host/src/host/metered_xdr.rs [hunk 4]
```diff
@@ -49,6 +51,13 @@ impl Host {
         })
     }
 
+    pub(crate) fn metered_hash_xdr(&self, obj: &impl WriteXdr) -> Result<[u8; 32], HostError> {
+        let mut buf = vec![];
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply soroban-env-host/src/host/metered_xdr.rs:55]: Sorry I wasn't clear. I wasn't advocating for you to use one of the other methods. My point was that you could update the other uses to `let mut buf = vec![];`. I'll leave it up to you though.

File: soroban-env-host/src/host.rs [hunk 5]
```diff
@@ -741,41 +744,69 @@ impl Host {
...
+        // without this check it would be possible to accidentally create a
+        // contract that never may be invoked (just by providing a bad hash).
+        if let ScContractCode::WasmRef(wasm_hash) = &contract_code {
+            let wasm_storage_key =
+                self.contract_code_ledger_key(wasm_hash.metered_clone(&self.0.budget)?);
+            if !self.0.storage.borrow_mut().has(&wasm_storage_key)? {
+                return Err(self.err_general("Contract code was not installed"));
+            }
+        }
+        self.store_contract_code(contract_code, new_contract_id, &storage_key)?;
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply soroban-env-host/src/host.rs]: Hmm source is confusing to me because I expect it to contain the source code, which isn't true here. Maybe you should leave it and open a cleanup issue? What do you think?

File: soroban-env-host/src/host/metered_xdr.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-host/src/host/metered_xdr.rs:55]: I actually don't see this in the affected parts of the PR. Also sometimes we need to have an explicit type provided... Anyway, my point is that I wouldn't worry much about consistency in this particular case unless there is an objectively better option (that could be cleaned up separately then).

File: soroban-env-host/src/host.rs [hunk 6]
```diff
@@ -704,11 +705,13 @@ impl Host {
             HostObject::BigInt(bi) => {
                 self.charge_budget(CostType::HostBigIntAllocCell, bi.bits() as u64)?;
             }
-            HostObject::ContractCode(cc) => {
-                if let ScContractCode::Wasm(c) = cc {
-                    self.charge_budget(CostType::HostContractCodeAllocCell, c.len() as u64)?;
+            HostObject::ContractCode(cc) => match cc {
+                ScContractCode::WasmRef(h) => {
```
Comment (leighmcculloch) [inline soroban-env-host/src/host.rs:709]: Fyi that clippy typically recommends using if let over match in situations where only a single case is required.

**Review by leighmcculloch — COMMENTED** [review]
Few questions inline, mostly relating to refactoring/changes that seem not to be core to the main thing being delivered in this change.

File: soroban-env-host/src/host.rs [hunk 7]
```diff
@@ -121,7 +117,7 @@ struct VmSlice {
     len: u32,
 }
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Default)]
```
[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-host/src/host.rs]: Good point, I've moved the `Default` impl to the test utils and made it return the 32x0 network id instead of an empty one.

File: soroban-env-host/src/host.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-host/src/host.rs]: The source is now consistently consistently refers to the implementation (either wasm or builtin). That's basically the 'class name/type' of the contract in OOP analogy. I realize this may be a bit confusing, but OTOH not sure what would be a better/less confusing name. Since changes would need to be done in XDR, let's maybe brainstorm this a bit and cleanup later.

**Comment by dmkozh** [comment]
PTAL

File: soroban-env-host/src/host.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-host/src/host.rs]: Opened https://github.com/stellar/rs-soroban-env/issues/575

File: soroban-env-host/src/host.rs [hunk 8]
```diff
@@ -704,11 +705,13 @@ impl Host {
...
-                if let ScContractCode::Wasm(c) = cc {
-                    self.charge_budget(CostType::HostContractCodeAllocCell, c.len() as u64)?;
+            HostObject::ContractCode(cc) => match cc {
+                ScContractCode::WasmRef(h) => {
+                    self.charge_budget(CostType::HostContractCodeAllocCell, h.0.len() as u64)?;
                 }
-            }
+                ScContractCode::Token => (),
+            },
+
```
[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-host/src/host.rs]: Nice, also I've needed to merge your changes anyway.

**Review by sisuresh — APPROVED** [review]
LGTM

File: soroban-env-host/src/host.rs (hunk 6 above)
Comment (dmkozh) [reply soroban-env-host/src/host.rs]: This is actually completely gone now, so no need for the check.

File: soroban-env-host/src/test/util.rs [hunk 9]
```diff
@@ -145,4 +132,39 @@ impl Host {
...
+        &self,
+        contract_wasm: &[u8],
+    ) -> Result<Object, HostError> {
+        self.set_source_account(generate_account_id());
```
[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-host/src/test/util.rs:151]: I'm marking this as resolved for now; we can come back to this discussion on the SDK PR and do the host changes if we find them to be 100% compatible with SDK.

File: soroban-env-host/src/test/util.rs (hunk 3 above)
Comment (dmkozh) [reply soroban-env-host/src/test/util.rs:147]: I'll mark this resolved for now to unlblock the non-test changes. We can come back to this discussion on the respective SDK PR.
